src/types.rs

```diff
@@ -462,17 +482,22 @@ impl Value {
     }
 
     fn resolve_union(self, schema: &UnionSchema) -> Result<Self, Error> {
-        let v = match self {
-            // Both are unions case.
-            Value::Union(v) => *v,
-            // Reader is a union, but writer is not.
-            v => v,
-        };
-        // Find the first match in the reader schema.
-        let (_, inner) = schema
-            .find_schema(&v)
-            .ok_or_else(|| SchemaResolutionError::new("Could not find matching type in union"))?;
-        v.resolve(inner)
+        match self {
```

Comment:
It looks like this code is missing the resolution when the writer value is not a union but reader is a union.

e.g: writer writes a `Long`, reader is `Union(Null, Long)`,  then `resolve_union` should output a `Union(1, Value::Long)`

See the "if reader's is a union, but writer's is not" section in https://avro.apache.org/docs/1.8.2/spec.html#Schema+Resolution

Reply:
The Standard relies on availability of both writer- and reader-schema for resolution.
So I believe the signature of the method should go as `pub(crate) fn resolve(reader_schema, writer_schema) -> Result<Value, Error>`.
The only usage of the method is `pub fn from_avro_datum<R: Read>`. I think it only would be right to remove the method from the public API.

Reply:
@flavray what do you think if something like that will be done for schema-resolution: https://github.com/RGafiyatullin/avro-rs/blob/rg-schema-resolution/src/schema_resolution.rs ?